apollo-federation/src/sources/connect/json_selection/lit_expr.rs

```diff
@@ -0,0 +1,522 @@
+// A LitExpr (short for LiteralExpression) is similar to a JSON value (or
+// serde_json::Value), with the addition of PathSelection as a possible leaf
+// value, so literal expressions passed to -> methods (via MethodArgs) can
+// capture both field and argument values and sub-paths, in addition to constant
+// JSON structures and values.
+
+use std::hash::Hash;
+
+use apollo_compiler::collections::IndexMap;
+use apollo_compiler::collections::IndexSet;
+use nom::branch::alt;
+use nom::bytes::complete::tag;
+use nom::character::complete::char;
+use nom::character::complete::one_of;
+use nom::combinator::map;
+use nom::combinator::opt;
+use nom::combinator::recognize;
+use nom::multi::many0;
+use nom::multi::many1;
+use nom::sequence::delimited;
+use nom::sequence::pair;
+use nom::sequence::preceded;
+use nom::sequence::tuple;
+use nom::IResult;
+
+use super::helpers::spaces_or_comments;
+use super::parser::parse_string_literal;
+use super::parser::Key;
+use super::parser::PathSelection;
+use super::CollectVarPaths;
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub enum LitExpr {
+    String(String),
+    Number(serde_json::Number),
+    Bool(bool),
+    Null,
+    Object(IndexMap<String, LitExpr>),
+    Array(Vec<LitExpr>),
+    Path(PathSelection),
+}
+
+impl LitExpr {
+    // LitExpr      ::= LitPrimitive | LitObject | LitArray | PathSelection
+    // LitPrimitive ::= LitString | LitNumber | "true" | "false" | "null"
+    pub fn parse(input: &str) -> IResult<&str, Self> {
+        tuple((
+            spaces_or_comments,
+            alt((
+                map(parse_string_literal, Self::String),
+                Self::parse_number,
+                map(tag("true"), |_| Self::Bool(true)),
+                map(tag("false"), |_| Self::Bool(false)),
+                map(tag("null"), |_| Self::Null),
```

Comment:
When would someone need to express `null` as a literal?

Reply:
https://github.com/apollographql/router/pull/5762#discussion_r1728001911